### PR TITLE
Fixed caching of Kover report tasks

### LINKED ISSUE #68: Kover report tasks not cacheable
When running a simple project twice with build cache enabled, we can see that the Kover report task is skipped:

image

which leads to no report being generated nor taken from a cache.
The way I would expect it working would be:
- First run stores the report into the build cache
- Second run does not skip the report task, just takes the report from the cache

### PR DESCRIPTION
**Pull request by shanshin**
Resolves #68

### CODE REVIEW

**Comment by qwwdfsad** [comment]
It definitely would be nice to have a test on it

File: src/main/kotlin/kotlinx/kover/api/KoverExtension.kt [hunk 1]
```diff
@@ -6,33 +6,39 @@ package kotlinx.kover.api
...
      * Specifies the version of JaCoCo dependency.
      */
     val jacocoEngineVersion: Property<String> = objects.property(String::class.java)
+        @Input get
 
     /**
      * Specifies whether the reports will be generated within 'check' task execution.
      */
     val generateReportOnCheck: Property<Boolean> = objects.property(Boolean::class.java)
+        @Input get
```
Comment (qwwdfsad) [inline src/main/kotlin/kotlinx/kover/api/KoverExtension.kt:41]: here and everywhere: it's enough to mark the field with the `@Input`/`@Output`/`@Nested` without the need to introduce getter separately, isn't it?

File: src/main/kotlin/kotlinx/kover/KoverPlugin.kt [hunk 2]
```diff
@@ -148,6 +150,24 @@ class KoverPlugin : Plugin<Project> {
...
+                .map { t -> t.extensions.getByType(KoverTaskExtension::class.java) }
+                .filter { e -> e.isEnabled }
+                .map { e -> e.smapFile.orNull }
+                /*
+                 Binary reports and SMAP files have same ordering for IntelliJ engine:
+                    * SMAP file is null if coverage engine is a JaCoCo by default - in this case property is unused
+                    * SMAP file not creates by JaCoCo - property is unused
+                    * test task have no sources - in this case binary report and SMAP file not exists
+                 */
+                .filter { f -> f?.exists() == true }
```
Comment (qwwdfsad) [inline src/main/kotlin/kotlinx/kover/KoverPlugin.kt:164]: super nit -- when working with nulls, `f?.exists() ?: true` can be more readab;e

File: src/main/kotlin/kotlinx/kover/engines/intellij/IntellijCoverage.kt [hunk 3]
```diff
@@ -95,8 +97,9 @@ private fun Writer.writeModuleReportJson(
     appendLine("""  "modules": [""")
     appendLine("""    { "reports": [ """)
 
+    val smapIterator = smapFiles.iterator()
     appendLine(binaryReportFiles.joinToString(",\n        ", "        ") { f ->
-        """{"ic": "${f.safePath()}", "smap": "${f.safePath()}.smap"}"""
+        """{"ic": "${f.safePath()}", "smap": "${smapIterator.next().safePath()}"}"""
```
Comment (qwwdfsad) [inline src/main/kotlin/kotlinx/kover/engines/intellij/IntellijCoverage.kt:102]: You can remove `smapIterator` and just use `smapIterator.first()`

**Review by qwwdfsad — APPROVED** [review]

File: src/main/kotlin/kotlinx/kover/api/KoverExtension.kt (hunk 1 above)
Comment (shanshin) [reply src/main/kotlin/kotlinx/kover/api/KoverExtension.kt:41]: Gradle searches for annotations only on getters.
Documentation to the annotation: `... Annotations on setters or just the field in Java are ignored.`

File: src/main/kotlin/kotlinx/kover/engines/intellij/IntellijCoverage.kt (hunk 3 above)
Comment (shanshin) [reply src/main/kotlin/kotlinx/kover/engines/intellij/IntellijCoverage.kt:102]: Unfortunately, `first` will not be enough - it is necessary to read all the elements of the iterator inside a loop strictly in order

Comment (qwwdfsad) [reply src/main/kotlin/kotlinx/kover/engines/intellij/IntellijCoverage.kt:102]: Oh, I see, it's captured in lambda.

**Comment by shanshin** [comment]
> It definitely would be nice to have a test on it

Tests added

File: src/main/kotlin/kotlinx/kover/api/KoverExtension.kt (hunk 1 above)
Comment (shanshin) [reply src/main/kotlin/kotlinx/kover/api/KoverExtension.kt:41]: Changed to `@get:Nested` etc

**Pull request merged**